### PR TITLE
Update react and react-dom to v16.x

### PR DESCRIPTION
**Pull request by latenssi**
- Improve eslint configuration
- Explicitly define supported browsers
- Make codebase compliant with React 16.x
- Update react and react-dom to 16.x
- Add TODO.md to track general todos

**babel-preset-env** allows to _babelify_ code to whatever env is defined (for example browser or Node 6). It also supports [browserslist](https://github.com/browserslist/browserslist) so we can define which browsers we want to support. This can be defined in a separate `.browserslistrc` file or as I have chosen to do in `package.json`. I think the `package.json` style of _browserslist_ config is clearer. I have also chosen to support _"last 2 versions"_, _"> 1%"_ and _"not dead"_. This decision is based on what I personally have used before and I am really **not sure** if it suits this project. I hope we can have an open discussion about what we should support.

**NOTE:** The PR is kind of big so I am in the works of separating it in to 2 PRs. So hold on with the reviews :)

### CODE REVIEW

**Comment by latenssi** [comment]
The PR splitting task was going to take too long so I'm going to scratch it this time. Sorry for the big PR. Most of the codeline changes are because of automatic linting. I will try to be more careful while creating commits in the future.

**Comment by latenssi** [comment]
The important changes regarding compatibility with React v16.x is the use of `create-react-class` and `prop-types` packages instead of build-in React modules. 

I added to the TODO list that we should also aim to drop the use of `create-react-class` package since today we can create "native" JS classes.

File: TODO.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+- ~~Update _React_ to v16.x~~
```
Comment (RichardLitt) [inline TODO.md]: Is there a better way to store TODOs? As issues, perhaps?

Comment (haadcode) [reply TODO.md]: Clarifying @RichardLitt's *slight hint* (😉), issues would be a great place for tracking the TODOs. That way others can contribute or keep track. Don't necessarily have to do one issue per task, could all be in one big "update all the things" issue, or broken down to smaller ones as you see fit.

File: src/components/App.js [hunk 2]
```diff
@@ -1,28 +1,23 @@
...
-import IpfsDaemonStore from 'stores/IpfsDaemonStore'
 import AppStateStore from 'stores/AppStateStore'
 import UserStore from 'stores/UserStore'
-import UserActions from 'actions/UserActions'
 import NetworkStore from 'stores/NetworkStore'
 import ChannelStore from 'stores/ChannelStore'
-import MessageStore from 'stores/MessageStore'
-import UsersStore from 'stores/UsersStore'
 import SettingsStore from 'stores/SettingsStore'
-import SwarmStore from 'stores/SwarmStore'
```
Comment (haadcode) [inline src/components/App.js]: Hah, awesome to remove so many unnecessary(?) deps from the component! 👍

File: TODO.md (hunk 1 above)
Comment (RichardLitt) [reply TODO.md]: :) Maybe a TODO file is the best! I don't know! I just know that I tend to use issues, and that the rest of the Orbit community seems to. :)

File: src/components/App.js [hunk 3]
```diff
@@ -84,62 +82,90 @@ var App = React.createClass({
...
+  _handleAppStateChange: function (state) {
+    let prefix = ''
 
-    if(!AppStateStore.state.hasFocus && AppStateStore.state.unreadMessages[AppStateStore.state.currentChannel] > 0)
-      suffix = `(${AppStateStore.state.unreadMessages[AppStateStore.state.currentChannel]})`
+    let suffix = ''
 
-    if(Object.keys(state.unreadMessages).length > 1 || (Object.keys(state.unreadMessages).length === 1 && !Object.keys(state.unreadMessages).includes(AppStateStore.state.currentChannel)))
+    if (
+      !AppStateStore.state.hasFocus &&
```
Comment (haadcode) [inline src/components/App.js]: Is the line break here due to linter? I personally find that a little harder to read, but if that's what the linter says, then let's go with that.

File: src/components/App.js [hunk 4]
```diff
@@ -84,62 +82,90 @@ var App = React.createClass({
...
 
-    if(Object.keys(state.mentions).length > 0)
-      prefix = '!'
+    if (Object.keys(state.mentions).length > 0) prefix = '!'
 
-    if(state.currentChannel) {
-      document.title = prefix + ' ' + AppStateStore.state.location + ' ' + suffix
-      this.goToLocation(state.currentChannel, views.Channel + encodeURIComponent(state.currentChannel))
+    if (state.currentChannel) {
+      document.title =
```
Comment (haadcode) [inline src/components/App.js]: I guess this is again due to linter? Would it fit on one line using string literals?
eg. `${prefix} ${AppStateStore.state.location} ${suffix}`

File: src/components/App.js [hunk 5]
```diff
@@ -152,103 +178,108 @@ var App = React.createClass({
...
   },
-  onLeaveChannel: function(channel) {
+  onLeaveChannel: function (channel) {
     const { user, networkName } = this.state
     const channelsKey = this._makeChannelsKey(user.name, networkName)
-    const channels = this._getSavedChannels(networkName, user.name).filter((c) => c.name !== channel)
-    if (channels.length === 0)
-      localStorage.removeItem(channelsKey)
-    else
+    const channels = this._getSavedChannels(networkName, user.name).filter(
```
Comment (haadcode) [inline src/components/App.js]: Stylistic nitpicking: I feel it'd be better to split this per function call, eg:
```javascript
const channels = this._getSavedChannels(networkName, user.name)
  .filter(c => c.name !== channel)
```

File: src/components/App.js [hunk 6]
```diff
@@ -152,103 +178,108 @@ var App = React.createClass({
...
+      c => c.name !== channel
+    )
+    if (channels.length === 0) localStorage.removeItem(channelsKey)
```
Comment (haadcode) [inline src/components/App.js]: Should prolly add `{}` around the `if` since they're used below for the `else`

File: src/components/App.js [hunk 7]
```diff
@@ -257,14 +288,16 @@ var App = React.createClass({
...
-        username={this.state.user ? this.state.user.name : ""}
+        username={this.state.user ? this.state.user.name : ''}
         requirePassword={this.state.requirePassword}
         theme={this.state.theme}
         left={this.state.leftSidePanel}
         networkName={this.state.networkName}
         joiningToChannel={this.state.joiningToChannel}
       />
-    ) : ""
+    ) : (
```
Comment (haadcode) [inline src/components/App.js]: Stylistic nitpicking: Why the `()` and line break (linter again?)? Could this be simple `''`?

File: src/components/NewMessageNotification.js [hunk 8]
```diff
@@ -1,34 +1,35 @@
...
           </div>
-        );
+        )
       } else {
         return (
           <div className="newMessagesBar" onClick={onClick}>
-            There are <span className="newMessagesNumber">{unreadMessages}</span> new messages
+            There are{' '}
+            <span className="newMessagesNumber">{unreadMessages}</span> new
+            messages
```
Comment (haadcode) [inline src/components/NewMessageNotification.js]: Something's off on these lines re. the line breaks, hard to read the full structure here.

File: src/components/generics/ListForm.js [hunk 9]
```diff
@@ -1,76 +1,80 @@
...
-  render() {
+  render () {
     const list = this.props.list.map((element, i) => {
       return (
-        <li key={element+i} onClick={() => {this.remove(element)} }>
+        <li
+          key={element + i}
+          onClick={() => {
+            this.remove(element)
+          }}
```
Comment (haadcode) [inline src/components/generics/ListForm.js]: These could all go on one line probably?

**Comment by haadcode** [comment]
@latenssi looking good! 👍 Left some comments on the style, but substance-wise the PR LGTM 👍 Happy to merge as is, and can leave the style stuff for later consideration.

File: TODO.md (hunk 1 above)
Comment (latenssi) [reply TODO.md]: Yea issues are probably the correct way here :) The "todo" file began its life as simple file for me to keep track but now that it is becoming larger I will swap to using proper issues.

File: src/components/App.js (hunk 2 above)
Comment (latenssi) [reply src/components/App.js]: Yup they were unused :)

**Comment by latenssi** [comment]
@haadcode You are correct that the styling issues are caused by linting. I am using [prettier ](https://github.com/prettier/prettier) with a plugin that makes it respect eslint rules so my code gets automatically formatted according to eslints and prettiers rules. This releases me from thinking about formatting and as I have accustomed to the tooling my coding has changed so that prettier and eslint can do their jobs nicely. Usually this works fine but when I do these kinds of sweeping changes it raises a couple of issues:
1. The convention will cause a lot of line changes
2. I might miss these kind of styling issues

I will refactor the 3 files that had these kind of styling issues to be more in line what kind of style I am used to code in. Would love to hear your feedback! Of course if you think this is a bad idea I will stop using the autoformatting :)

**Comment by haadcode** [comment]
@latenssi no no, you're absolutely right, since we're doing all this work, we should finally start using some linting here! 👍 I think the confusion was that in orbitdb repos we're using (planning to use) eslint directly which seems to have some difference to prettier's formatting. But like said, this is totally fine as it is and happy to use whatever you feel is the best for the job. All good! 😃

**Pull request merged**